activerecord/lib/active_record/persistence.rb

```diff
@@ -100,10 +100,11 @@ def instantiate(attributes, column_types = {}, &block)
       def update(id = :all, attributes)
         if id.is_a?(Array)
           id.map.with_index { |one_id, idx|
-            object = find_by(primary_key => one_id)
-            object.update(attributes[idx]) if object
+            [ find(one_id), idx ]
```

Comment:
I think this one could also `find(id).each.with_index { |object, idx| object.update(attributes[idx]) }`

No real difference, but saves the explicit double-map.

Reply:
The reason for using double-map is that `ids.flatten.compact.uniq` is executed in `find_with_ids`.
So if `update([nil, 1, 2], [{}, { "content" => "1 updated" }, { "content" => "2 updated" }])` is called, the record 1 will be updated with incorrect the first `{}`.

https://github.com/rails/rails/blob/d041a1dcbaced9f44ed976a48312c9f492fffe5e/activerecord/lib/active_record/relation/finder_methods.rb#L429-L435

Reply:
Did we lose #20338 at some point? :confused:

Reply:
Oh, that never specified duplicates/nils keep their place. Hmm.. I think it should, because this is exactly the sort of use case you want that for.

But then again, we'd still have the explicit-`order`-overrides-id-order issue. 🤦🏻

:+1: (though I think you can still move the .with_index to the second map)

(and please add a test for the above [if there's not already], because I just know I'll come along and try to be clever later :sweat_smile:)